src/realm/util/compression.cpp

```diff
@@ -311,23 +708,102 @@ void compression::allocate_and_compress(CompressMemoryArena& compress_memory_are
-                        n = std::numeric_limits<std::size_t>::max();
-                }
-                compress_memory_arena.resize(n); // Throws
+                grow_arena(compress_memory_arena); // Throws
                 continue;
             }
-            throw std::system_error(ec);
+            return ec;
         }
         break;
     }
     compressed_buf.resize(compressed_size);
+    return std::error_code{};
+}
+
+void compression::allocate_and_compress_with_header(CompressMemoryArena& arena, Span<const char> uncompressed,
+                                                    util::AppendBuffer<char>& compressed)
+{
+    compressed.resize(uncompressed.size() + 10);
+    uint64_t size = uncompressed.size();
+    std::memcpy(compressed.data(), &size, sizeof(size));
+
+    size_t compressed_size = 0;
+    // zlib is ineffective for very small sizes. Measured results indicate that
+    // it only manages to compress at all past 100 bytes and the compression
+    // ratio becomes interesting around 200 bytes.
+    while (uncompressed.size() > 256) {
+        init_arena(arena);
+        const int compression_level = 1;
+        auto ec = compress(uncompressed, Span(compressed).sub_span(8), compressed_size, compression_level, &arena);
+        if (ec == error::compress_buffer_too_small) {
+            // Compressed result was larger than uncompressed, so just store the
+            // uncompressed
+            compressed_size = 0;
+            break;
+        }
+        if (ec == compression::error::out_of_memory) {
+            grow_arena(arena); // Throws
+            continue;
+        }
+        if (ec) {
+            throw std::system_error(ec);
+        }
+        record_compression_result(uncompressed.size(), compressed_size + 8);
+        REALM_ASSERT(compressed_size);
+        compressed.resize(compressed_size + 8);
+        break;
+    }
+
+    // If compression made it grow or it was too small to compress then copy
+    // the source over uncompressed
+    if (!compressed_size) {
+        record_compression_result(uncompressed.size(), uncompressed.size() + 10);
+        compressed.data()[8] = static_cast<char>(Algorithm::None);
+        compressed.data()[9] = 0; // Window size
+        std::memcpy(compressed.data() + 10, uncompressed.data(), uncompressed.size());
+    }
+}
+
+util::AppendBuffer<char> compression::allocate_and_compress_with_header(Span<const char> uncompressed_buf)
+{
+    util::compression::CompressMemoryArena arena;
+    util::AppendBuffer<char> compressed;
+    allocate_and_compress_with_header(arena, uncompressed_buf, compressed);
+    return compressed;
+}
+
+std::unique_ptr<NoCopyInputStream> compression::decompress_input_stream(NoCopyInputStream& source, size_t& total_size)
+{
+    auto first_block = source.next_block();
+    uint64_t size = 0;
+    for (int i = 0; i < 8; ++i) {
+        size += uint64_t(read_byte(source, first_block)) << (8 * i);
```

Comment:
Why do we store this above by std::memcpy'ing uint64_t's around and do this loop down here?

Reply:
All the uses of memcpy() on ints should be gone in the final version.

The loop is needed here because the int can straddle multiple blocks, while when compressing we're always writing to a single buffer.